### ISSUE TITLE
Additional WAYF entries per IdP Endpoint with dedicated name, logo, keywords

### ISSUE THREAD

**Issue by govroam** [OpenConext/OpenConext-engineblock#1338]
Currently, an IdP endpoint can be displayed in the WAYF only once under it's human readable name with logo and accompanied by keywords.
In the context of govconext, multiple organisations (IdP's) can be facilitated over one SAML relation that each have their own name, schacHomeOrganization FQDN's, log and keywords. For example, a Shared Service Center connected to govconext may manage the user accounts for multiple municipalities that each have their own characteristics. Assuming those municipalities could be characterized by the term 'IdP subtenants', the end users of these 'IdP subtenants' may not be aware of the fact that they need to look for the Shared Service Center in the WAYF. We currently solve this through keywords, but have received multiple requests to mention those 'IdP subtenant' organisations behind the Shared Service Center individually in the WAYF.

This request is merely a cosmetical addition to the WAYF for those organisations. As far as we are concerned, policies and other characteristic of the SAML endpoint (of the Shared Service Center in this case) may apply to all 'underlying' 'IdP subtenants'. There is no need that these 'IdP subtenants' have their own dedicated organisational entry in Manage with corresponding parameters.

**Comment by baszoetekouw** [comment]
There are some things that we need to decide before we start developing this:

- How to implement this in Manage (see https://github.com/OpenConext/OpenConext-manage/issues/457)
- What should the pushed metadata for these entities look like?
- The WAYF-implementation is probably trivial (simple insert multiple entries in the WAYF that result in the same entityid being selected).  But what should happen if the logo/displayname is needed later in the flow?  For example, in error messages, consent screen, etc?  Should we show the "main" logo for the IdP, or should we show the one that is selected by the user?

**Comment by tvdijen** [comment]
We have IdP's that are the home of multiple independent top-level organisations, so for me a different display name is a solid requirement for this to become useful. Logos are irrelevant to us, but would make sense to show the one from the sub-org.

**Comment by baszoetekouw** [comment]
Once https://github.com/OpenConext/OpenConext-manage/issues/457 is complete, we can start implementing this in EB, I think, as that clears up the first two questions in https://github.com/OpenConext/OpenConext-engineblock/issues/1338#issuecomment-2535024476

Regarding this issues:
> The WAYF-implementation is probably trivial (simple insert multiple entries in the WAYF that result in the same entityid being selected). But what should happen if the logo/displayname is needed later in the flow? For example, in error messages, consent screen, etc? Should we show the "main" logo for the IdP, or should we show the one that is selected by the user?

We should probably store the index of the chosen "subentity" in the session so that we can look up the correct details if we need them later on during AuthnResponse processing.

**Comment by johanib** [comment]
Discussed with @baszoetekouw 

* `IDP Subtenants` are shows as actual IDP's on the wayf.
* No extra consent changes are needed, keep as is (once per actual IDP)
* Pay attention to backwards compatibility when implementing the metadata push changes, because red/blue release strategy.

**Comment by baszoetekouw** [comment]
@tvdijen to get an idea, what is the maximum number of logos/names you need for a single entity?

**Comment by baszoetekouw** [comment]
I've asked @oharsta to implement the required changes in Manage in https://github.com/OpenConext/OpenConext-manage/issues/457
Once that is finished, we'll have an input example for Engineblock.

**Comment by oharsta** [comment]
This is the proposed format for all new discovery elements in the `metadata` section of an `IdentityProvider`. The change is backwards compatible and `logo` and `keywords` are still present in the push. 
```
        "discoveries": [
          {
            "keywords_en": "keywords, (test)",
            "keywords_nl": "keywords, nl",
            "logo_height": "96",
            "logo_url": "https://static.test2.surfconext.nl/media/conext_logo.png",
            "logo_width": "96",
            "name_en": "IdP test 0",
            "name_nl": "IdP test 0"
          },
          {
            "keywords_en": "keywords 1 en",
            "keywords_nl": "keywords 1 nl",
            "logo_url": "https://logo1.url",
            "name_en": "disco name en 1",
            "name_nl": "disco name nl 1"
          },
          {
            "keywords_en": "key2en",
            "keywords_nl": "key2nl",
            "logo_url": "https://logo2.url",
            "logo_width": "120",
            "name_en": "disco name en 2",
            "name_nl": "disco name nl 2"
          },
          {
            "name_en": "dsname 3"
          }
        ],
```
Not each individual discovery element has to be complete. it is either up to EB to determine if an element is detailed enough to be displayed or I'll add an extra validation in Manage to check and enforce the completeness. @baszoetekouw Your call.

Note this is just a proposal and it is very easy to change the format if necessary. I do value the backward compatibility, so that will not change.

**Comment by baszoetekouw** [comment]
Let's implement the checks on both sided: Manage should try and send only complete records, and EB should refuse to accept incomplete ones.

**Comment by tvdijen** [comment]
Logo shouldn't be a requirement..  Just `name_en`

**Comment by johanib** [comment]
What happens when no `logo_url` is provided?
The 'tenant' will show no logo? Or the logo of the idp?

**Comment by baszoetekouw** [comment]
> What happens when no `logo_url` is provided? Should the 'tenant' show no logo? Or the logo of the idp?

The current behaviour is like this:
![Image](https://github.com/user-attachments/assets/3b0cc2c7-61be-473a-b398-f359063aaf69)

I see no reason to change that.  So, for each "sub-idp", if no logo is present, show the generic "no-logo" image.

**Comment by oharsta** [comment]
New [9.0.1-SNAPSHOT of Manage](https://manage.test2.surfconext.nl/) and the main branch of [OpenConext-Deploy](https://github.com/OpenConext/OpenConext-deploy/blob/main/roles/manage/templates/metadata_configuration/saml20_idp.schema.json.j2#L445) have support for multiple names, logo's and keywords. See https://github.com/OpenConext/OpenConext-manage/issues/457

**Comment by johanib** [comment]
@baszoetekouw 
When you have used an IdP in the past, on the next login, that IdP is pre-selected.
Currently, with the new discoveries, two IdP's are shown, as its technically the same IdP. (One IdP & one discovery)

I think we should we update this logic to base the history on `entityId` + `discovery hash` instead of just the `entityId`?
Do you agree, or is this out of scope?

![Image](https://github.com/user-attachments/assets/7984cd30-f6a0-416e-ae83-d585aa473ea3)

**Comment by tvdijen** [comment]
Should be in scope, because it would kill the user experience if they see this selection screen on every continuous login.
It's probably also a good idea to also test the 'remember choice' feature in this regard.
